include/yaclib/coro/detail/await_on_awaiter.hpp

```diff
@@ -0,0 +1,132 @@
+      if (this->SubEqual(1)) {
+        YACLIB_ASSERT(_job != nullptr);
+        _executor.Submit(*_job);
+      }
+    }
+    return Noop<SymmetricTransfer>();
+  }
+  [[nodiscard]] InlineCore* Here(InlineCore& caller) noexcept final {
+    return Impl<false>(caller);
+  }
+#if YACLIB_SYMMETRIC_TRANSFER != 0
+  [[nodiscard]] yaclib_std::coroutine_handle<> Next(InlineCore& caller) noexcept final {
+    return Impl<true>(caller);
+  }
+#endif
+  BaseCore* _job{nullptr};
+  IExecutor& _executor;
+};
+
+template <bool Single>
+struct [[nodiscard]] AwaitOnAwaiter final {
+  explicit AwaitOnAwaiter(IExecutor& e, BaseCore& caller) noexcept : _executor{e}, _caller{caller}, _event{1, e} {
+  }
+
+  YACLIB_INLINE bool await_ready() const noexcept {
+    return false;
+  }
+
+  template <typename Promise>
+  YACLIB_INLINE void await_suspend(yaclib_std::coroutine_handle<Promise> handle) noexcept {
+    BaseCore& core = handle.promise();
+    core._executor = &_executor;
+    _event.SetJob(&core);
+
+    if (!_caller.SetCallback(_event)) {
+      _executor.Submit(core);
+    }
+  }
+
+  constexpr void await_resume() const noexcept {
+  }
+
+ private:
+  IExecutor& _executor;
+  BaseCore& _caller;
+  AwaitOnEvent<Single> _event;
+};
+
+template <>
+class [[nodiscard]] AwaitOnAwaiter<false> final {
+ public:
+  template <typename... Cores>
+  explicit AwaitOnAwaiter(IExecutor& e, Cores&... cores) noexcept;
+
+  template <typename It>
+  explicit AwaitOnAwaiter(IExecutor& e, It it, std::size_t count) noexcept;
+
+  YACLIB_INLINE bool await_ready() const noexcept {
+    return false;
+  }
+
+  template <typename Promise>
+  YACLIB_INLINE void await_suspend(yaclib_std::coroutine_handle<Promise> handle) noexcept {
+    BaseCore& core = handle.promise();
+    core._executor = &_executor;
+    _event.SetJob(&core);
+
+    if (_event.SubEqual(1)) {
+      _executor.Submit(core);
+    }
+  }
+
+  constexpr void await_resume() const noexcept {
+  }
+
+ private:
+  IExecutor& _executor;
+  AwaitOnEvent<false> _event;
+};
+
+template <typename... Cores>
+AwaitOnAwaiter<false>::AwaitOnAwaiter(IExecutor& e, Cores&... cores) noexcept
+  : _executor{e}, _event{sizeof...(cores) + 1, e} {
+  static_assert(sizeof...(cores) >= 2, "Number of futures must be at least two");
+  static_assert((... && std::is_same_v<BaseCore, Cores>), "Futures must be Future in Wait function");
+  const auto wait_count = (... + static_cast<std::size_t>(cores.SetCallback(_event)));
+  _event.count.fetch_sub(sizeof...(cores) - wait_count, std::memory_order_relaxed);
+}
+
+template <typename It>
+AwaitOnAwaiter<false>::AwaitOnAwaiter(IExecutor& e, It it, std::size_t count) noexcept
+  : _executor{e}, _event{count + 1, e} {
+  std::size_t wait_count = 0;
+  for (std::size_t i = 0; i != count; ++i) {
+    wait_count += static_cast<std::size_t>(it->GetCore()->SetCallback(_event));
+    ++it;
+  }
+  _event.count.fetch_sub(count - wait_count, std::memory_order_relaxed);
+}
+
+}  // namespace yaclib::detail
```

Comment:
Please add empty line in the end of file.
We have it in our precommit hook, but it's probably doesn't work on Windows